File: apps/agni-server/src/core/interactions/patrimony/addSnapshotToPatrimony.ts
import { IUsecase } from "../interfaces";
import { ResourceNotFoundError } from "@core/errors/resournceNotFoundError";
import { PatrimonySnapshot } from "@core/domains/entities/patrimonySnapshot";
import { CreatedDto } from "@core/dto/base";
import { GetUID } from "@core/adapters/libs";
import { mapperTransactionStatus } from "@core/domains/constants";
import Repository from "@core/adapters/repository";
import { Patrimony } from "@core/domains/entities/patrimony";

export type RequestAddSnapshotPatrimony = {
    patrimonyId: string
    balance: number
    status: string
    date: Date
}

export class AddSnapshotPatrimonyUseCase implements IUsecase<RequestAddSnapshotPatrimony, CreatedDto> {
    private patrimonyRepo: Repository<Patrimony>
    private snapshotRepo: Repository<PatrimonySnapshot>

    constructor(patrimonyRepo: Repository<Patrimony>, snapshotRepo: Repository<PatrimonySnapshot>) {
        this.patrimonyRepo = patrimonyRepo
        this.snapshotRepo = snapshotRepo
    }

    async execute(request: RequestAddSnapshotPatrimony): Promise<CreatedDto> {
        if (!(await this.patrimonyRepo.get(request.patrimonyId)))
            throw new ResourceNotFoundError("PATRIMONY_NO_FOUND")

        const snapshot = new PatrimonySnapshot(GetUID(), request.patrimonyId, request.balance, 
            mapperTransactionStatus(request.status), request.date)

        await this.snapshotRepo.create(snapshot)

        return { newId: snapshot.getId() }
    }
}

File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/core/usecases/patrimonies/snapshots/AddSnapshotToPatrimony.kt
package dev.auguste.agni_api.core.usecases.patrimonies.snapshots

import dev.auguste.agni_api.core.adapters.repositories.IRepository
import dev.auguste.agni_api.core.entities.Patrimony
import dev.auguste.agni_api.core.entities.PatrimonySnapshot
import dev.auguste.agni_api.core.usecases.CreatedOutput
import dev.auguste.agni_api.core.usecases.interfaces.IUseCase
import dev.auguste.agni_api.core.usecases.patrimonies.snapshots.dto.AddSnapshotToPatrimonyInput

class AddSnapshotToPatrimony(
    val patrimonyRepo: IRepository<Patrimony>,
    val snapshotPatrimonyRepo: IRepository<PatrimonySnapshot>
): IUseCase<AddSnapshotToPatrimonyInput, CreatedOutput> {

    override fun execAsync(input: AddSnapshotToPatrimonyInput): CreatedOutput {
        patrimonyRepo.get(input.patrimonyId) ?: throw Error("Patrimony not found")

        val snapShot = PatrimonySnapshot(
            patrimonyId = input.patrimonyId,
            currentBalanceObserved = input.balance,
            date = input.date,
            status = input.status
        )

        snapshotPatrimonyRepo.create(snapShot)

        return CreatedOutput(snapShot.id)
    }
}